pkg/protocol/engine/commitmentfilter/accountsfilter/commitmentfilter.go

```diff
@@ -0,0 +1,210 @@
+	defer c.futureBlocksMutex.RUnlock()
+
+	// If we are not able to load the commitment for the block, it means we haven't committed this slot yet.
+	if _, err := c.commitmentFunc(block.ProtocolBlock().SlotCommitmentID.Index()); err != nil {
+		lo.Return1(c.futureBlocks.Get(block.ProtocolBlock().SlotCommitmentID.Index(), true).GetOrCreate(block.ProtocolBlock().SlotCommitmentID, func() ds.Set[*model.Block] {
+			return ds.NewSet[*model.Block]()
+		})).Add(block)
+
+		return true
+	}
+
+	return false
+}
+
+func (c *CommitmentFilter) ProcessPreFilteredBlock(block *model.Block) {
+	// TODO: redo optimisations to mark future block rather than checking each time.
+	if c.isFutureBlock(block) {
+		return
+	}
+
+	// check if the account exists in the specified slot.
+	accountData, exists, err := c.accountRetrieveFunc(block.ProtocolBlock().IssuerID, block.ProtocolBlock().SlotCommitmentID.Index())
+	if err != nil {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ierrors.Wrapf(err, "could not retrieve account information for block issuer %s", block.ProtocolBlock().IssuerID),
+		})
+
+		return
+	}
+	if !exists {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ierrors.Wrapf(err, "block issuer account %s does not exist in slot commitment %s", block.ProtocolBlock().IssuerID, block.ProtocolBlock().SlotCommitmentID.Index()),
+		})
+
+		return
+	}
+
+	// Check that the issuer key is valid for this block issuer and that the signature is valid
+	edSig, isEdSig := block.ProtocolBlock().Signature.(*iotago.Ed25519Signature)
+	if !isEdSig {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ierrors.Wrapf(ErrInvalidSignature, "only ed2519 signatures supported, got %s", block.ProtocolBlock().Signature.Type()),
+		})
+
+		return
+	}
+	if !accountData.PubKeys.Has(edSig.PublicKey) {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ierrors.Wrapf(ErrInvalidSignature, "block issuer account %s does not have public key %s in slot %d", block.ProtocolBlock().IssuerID, edSig.PublicKey, block.ProtocolBlock().SlotCommitmentID.Index()),
+		})
+
+		return
+	}
+	signingMessage, err := block.ProtocolBlock().SigningMessage(c.apiProvider.LatestAPI())
+	hiveEd25519.Verify(edSig.PublicKey[:], signingMessage, edSig.Signature[:])
+	if err != nil {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ierrors.Wrapf(ErrInvalidSignature, "error: %s", err.Error()),
+		})
+
+		return
+	}
+	if !hiveEd25519.Verify(edSig.PublicKey[:], signingMessage, edSig.Signature[:]) {
+		c.events.BlockFiltered.Trigger(&commitmentfilter.BlockFilteredEvent{
+			Block:  block,
+			Reason: ErrInvalidSignature,
+		})
+
+		return
+	}
+
+	// Check that the issuer of this block has non-negative block issuance credit
+	if accountData.Credits.Value < 0 {
```

Comment:
This should be checked much earlier, at least before the signature check because it is much cheaper than the signature check.

Reply:
Moved to right after loading the account data.